### PR TITLE
cudaPackages.cudnn: add missing nvrtc runtime dependency

### LINKED ISSUE #461334: pythonPackages.torch: cudnn runtime compilation fails - libnvrtc not in runpath
### Nixpkgs version

- Unstable (25.11)

### Describe the bug

in torch 2.9, cudnn can't load libnvrtc. this is supposed to work via this injected runpath https://github.com/NixOS/nixpkgs/blob/9da7f1cf7f8a6e2a7cb3001b048546c92a8258b4/pkgs/development/python-modules/torch/source/default.nix#L726

but somewhere between torch 2.8 and 2.9 this broke.
this is related to https://github.com/NixOS/nixpkgs/issues/296179

### Steps to reproduce

on nixpkgs with torch 2.9.0, and with torch 2.8, try running the following script with Python + Torch + CUDA on a fancy GPU like an H100.

```

import torch
import os
import ctypes

print(f"PyTorch version: {torch.__version__}")
print(f"CUDA available: {torch.cuda.is_available()}")
print(f"cuDNN version: {torch.backends.cudnn.version()}")
print(f"\nLD_LIBRARY_PATH: {os.environ.get('LD_LIBRARY_PATH', 'NOT SET')}")

device = torch.device("cuda:0")

# seems to trigger runtime compilation
batch, heads, seq_len, head_dim = 2, 32, 4096, 128

q = torch.randn(batch, heads, seq_len, head_dim, device=device, dtype=torch.bfloat16, requires_grad=True)
k = torch.randn(batch, heads, seq_len, head_dim, device=device, dtype=torch.bfloat16, requires_grad=True)
v = torch.randn(batch, heads, seq_len, head_dim, device=device, dtype=torch.bfloat16, requires_grad=True)

# causal mask + large seq len on an H100 triggers cudnn runtime compilation
try:
    output = torch.nn.functional.scaled_dot_product_attention(
        q, k, v,
        attn_mask=None,
        dropout_p=0.0,
        is_causal=True,  # triggers torch to compile a fused kernel
        scale=None
    )
    loss = output.sum()
    loss.backward()
    print("compiled fused kernel!")
except RuntimeError as e:
    print(f"failed to do sdpa: {e}")
    if "libnvrtc" in str(e):
        print("\n yayyyy we  reproduced the issue!")
```

### Expected behaviour

we see "compiled fused kernel!" as in torch 2.8

### Screenshots

_No response_

### Relevant log output

```console
Torch 2.8:
(nix:nix-shell-env) ari@dell-h100-nj4-2:~/src/psyche$ python test.py
PyTorch version: 2.8.0
CUDA available: True
cuDNN version: 91100

LD_LIBRARY_PATH: /mnt/weka/home/ari/.cache/nix-gl-host/229e49f198496607fb9a0a2eb54ac0f6f1e2590dced32f7aa4816a2927d48c6c/glx:/mnt/weka/home/ari/.cache/nix-gl-host/229e49f198496607fb9a0a2eb54ac0f6f1e2590dced32f7aa4816a2927d48c6c/cuda:/mnt/weka/home/ari/.cache/nix-gl-host/229e49f198496607fb9a0a2eb54ac0f6f1e2590dced32f7aa4816a2927d48c6c/egl:/mnt/weka/home/ari/.cache/nix-gl-host/3076b0246cb1468199a8444860ebaebe5ec5081f85098057a9f4d5b40c3de738/glx:/mnt/weka/home/ari/.cache/nix-gl-host/3076b0246cb1468199a8444860ebaebe5ec5081f85098057a9f4d5b40c3de738/cuda:/mnt/weka/home/ari/.cache/nix-gl-host/3076b0246cb1468199a8444860ebaebe5ec5081f85098057a9f4d5b40c3de738/egl:/mnt/weka/home/ari/.cache/nix-gl-host/4b2f4473e24574e8f7e12b6eaeb2210f9987d5836ca53404b80edb821c973b37/glx:/mnt/weka/home/ari/.cache/nix-gl-host/4b2f4473e24574e8f7e12b6eaeb2210f9987d5836ca53404b80edb821c973b37/cuda:/mnt/weka/home/ari/.cache/nix-gl-host/4b2f4473e24574e8f7e12b6eaeb2210f9987d5836ca53404b80edb821c973b37/egl:/mnt/weka/home/ari/.cache/nix-gl-host/5289cd0f221ee1de420da77318a67f1263ad9bd146c64a9b41544e06ed6678f8/glx:/mnt/weka/home/ari/.cache/nix-gl-host/5289cd0f221ee1de420da77318a67f1263ad9bd146c64a9b41544e06ed6678f8/cuda:/mnt/weka/home/ari/.cache/nix-gl-host/5289cd0f221ee1de420da77318a67f1263ad9bd146c64a9b41544e06ed6678f8/egl:/nix/store/r09vfy9hvswflrzmr3jgv1pdb7dwb8ym-rdma-core-59.0/lib

compiled fused kernel!

====================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

Torch 2.9:
(nix:nix-shell-env) ari@dell-h100-nj4-2:~/src/psyche$ python test.py
PyTorch version: 2.9.0
CUDA available: True
cuDNN version: 91300

LD_LIBRARY_PATH: /mnt/weka/home/ari/.cache/nix-gl-host/229e49f198496607fb9a0a2eb54ac0f6f1e2590dced32f7aa4816a2927d48c6c/glx:/mnt/weka/home/ari/.cache/nix-gl-host/229e49f198496607fb9a0a2eb54ac0f6f1e2590dced32f7aa4816a2927d48c6c/cuda:/mnt/weka/home/ari/.cache/nix-gl-host/229e49f198496607fb9a0a2eb54ac0f6f1e2590dced32f7aa4816a2927d48c6c/egl:/mnt/weka/home/ari/.cache/nix-gl-host/3076b0246cb1468199a8444860ebaebe5ec5081f85098057a9f4d5b40c3de738/glx:/mnt/weka/home/ari/.cache/nix-gl-host/3076b0246cb1468199a8444860ebaebe5ec5081f85098057a9f4d5b40c3de738/cuda:/mnt/weka/home/ari/.cache/nix-gl-host/3076b0246cb1468199a8444860ebaebe5ec5081f85098057a9f4d5b40c3de738/egl:/mnt/weka/home/ari/.cache/nix-gl-host/4b2f4473e24574e8f7e12b6eaeb2210f9987d5836ca53404b80edb821c973b37/glx:/mnt/weka/home/ari/.cache/nix-gl-host/4b2f4473e24574e8f7e12b6eaeb2210f9987d5836ca53404b80edb821c973b37/cuda:/mnt/weka/home/ari/.cache/nix-gl-host/4b2f4473e24574e8f7e12b6eaeb2210f9987d5836ca53404b80edb821c973b37/egl:/mnt/weka/home/ari/.cache/nix-gl-host/5289cd0f221ee1de420da77318a67f1263ad9bd146c64a9b41544e06ed6678f8/glx:/mnt/weka/home/ari/.cache/nix-gl-host/5289cd0f221ee1de420da77318a67f1263ad9bd146c64a9b41544e06ed6678f8/cuda:/mnt/weka/home/ari/.cache/nix-gl-host/5289cd0f221ee1de420da77318a67f1263ad9bd146c64a9b41544e06ed6678f8/egl:/nix/store/gcysxwq43ssx933b31d236c4w5gpl5l3-cuda12.8-cuda_nvrtc-12.8.93/lib:/nix/store/fbzrrpxa4i7781bwzn82dsid2scdvdjp-rdma-core-59.0/lib
failed to do sdpa: cuDNN Frontend error: No valid engine configs for Matmul_MUL_GEN_INDEX_GEN_INDEX_CMP_GE_BINARY_SELECT_Reduction_SUB_EXP_Reduction_LOG_ADD_DIV_Matmul_
{"engineId":2,"smVersion":900,"knobChoices":{"CUDNN_KNOB_TYPE_KERNEL_CFG":7}}
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: Could not open libnvrtc at: this->libnvrtc == nullptr
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: compiler.load()
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: rtk(kernelNumRunning)->loadDLL()
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: status == CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: ptr->isSupported()
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: finalize_internal()

{"engineId":1,"smVersion":900,"knobChoices":{"CUDNN_KNOB_TYPE_KERNEL_CFG":3}}
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: Could not open libnvrtc at: this->libnvrtc == nullptr
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: compiler.load()
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: rtk(kernelNumRunning)->loadDLL()
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: status == CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: ptr->isSupported()
Warning: CUDNN_STATUS_NOT_SUPPORTED_RUNTIME_PREREQUISITE_MISSING; Reason: finalize_internal()



 yayyyy we  reproduced the issue!
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.0-119-generic, Ubuntu, 22.04.4 LTS (Jammy Jellyfish), nobuild`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.6`
 - nixpkgs: `/nix/store/yff8cp4mrjrf71i1g9llxqpip8a6sfyj-source`

### Notify maintainers

@GaetanLepage @teh @thoughtpolice @tscholak 
---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)

### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+bug%22+-label%3A%226.topic%3A+darwin%22+-label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by ethernet8023**
cudnn uses nvrtc at runtime. this dependency wasn't specified anywhere, so this would fail unless nvrtc was injected into your runpath somewhere else. this is only triggered for specific operations on specific hardware.

This closes https://github.com/NixOS/nixpkgs/issues/461334

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by arilotter** [comment]
> > cudnn dlopens nvrtc at runtime.
> 
> This might be true, but at least that doesn't happen in your logs: https://gist.github.com/arilotter/0993568341df8df1cc9aa422b6735cfd#file-gistfile1-txt-L2130-L2144
> 
> H'm, and you've already confirmed this change solves your problem?

yeah you're right. huh. I'm even more confused now. 
i can confirm the change does solve my issue, and my repro script from the linked issue passes after this change. 

I'm so confused 😭

**Review by GaetanLepage — APPROVED** [review]

**Review by ConnorBaker — APPROVED** [review]
